atc/engine/check_delegate.go

```diff
@@ -124,12 +124,22 @@ func (d *checkDelegate) WaitToRun(ctx context.Context, scope db.ResourceConfigSc
 	runAt := end.Add(interval)
 
 	shouldRun := false
-	if d.build.IsManuallyTriggered() {
-		// ignore interval for manually triggered builds
-		shouldRun = true
-	} else if !d.clock.Now().Before(runAt) {
-		// run if we're past the last check end time
+	if d.build.Name() != db.CheckBuildName {
```

Comment:
I don't think we should overload the `.Name()` in this way - that's supposed to be for display purposes only. Could you explain the issue in the `watsjs` suite that this fixed?

Reply:
A failed `wats` build: https://ci.concourse-ci.org/teams/contributor/pipelines/prs/jobs/watsjs/builds/529?vars.branch=%22master%22. I couldn't exactly tell  why it failed, as I had no way to access ATC logs for the failed build. But when I debugged the failure, I found that, if I reverted my change from `check_delegate.go`, `wats` test would pass. Then I further realized that, for `task` embedded check, `check-delegate` actually holds the current running build rather than a check build. In any case, comparing current running build's creating with the resource's last check time doesn't make much sense.

So I added the logic of checking if `check-delegate` holds a normal build or a check build. If that's a normal build, that indicates a step embedded check, in that case, always run the check.

If `build.Name()` should not be used to distinguish normal build and check build, then other way I can see if is to check `build.ResourceId()`. Only check build returns a resource id.

Reply:
FWIW it's fairly easy to run the `watsjs` suite directly; there are [instructions in the contributing docs](https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#running-the-web-acceptance-tests-webwats).

I ran the tests locally without this fix and saw the same failure. What's actually happening is that the test aborts the build, which interrupts the first `check` step that runs to find the task's image's version. Here's a screenshot:

![image](https://user-images.githubusercontent.com/1880/115740574-f48afc00-a35c-11eb-8413-0d7d2e0afa6c.png)

All of these `image`s refer to the same resource. Because the first check step bumped the `LastCheckStartTime`, the later check steps decide not to run. But because the original check was interrupted, it didn't actually save any versions, and there isn't a version for them to return.

So I don't think this is caused by it being a check build vs. a 'regular' build - it's more because they're deciding not to run even when there has never been a previous successful run.

I think there are a couple possible fixes that would be a bit more accuratethan checking the build model:

* Check if the scope has a latest version and return `true` if it doesn't. This is technically correct, but it would cost an additional query.
* Instead of checking `build.ResourceID()`, check for `plan.Resource != ""`. This is a tiny change but makes it a bit more precise IMO, since it'll only affect the step we're actually trying to optimize, rather than all `check` steps in the build's plan. A resource checking build could contain a `check` for its resource type's image, too, and maybe more `check` steps in the future when we implement notifications. The check delegate already has similar logic [in other parts of `WaitToRun`](https://github.com/concourse/concourse/blob/0af59c6a895b1f6acfeaade85d359699ddc29e8b/atc/engine/check_delegate.go#L80) so it's at least a bit more consistent.

The one downside I can think of with the second approach is that it doesn't technically fix the core issue: if we ever have a build plan that has a `check` for a pipeline resource followed by a `get` of its version it's possible to end up with the same error if the resource never successfully checked before. No build plans currently do this, and I don't think there's anything relevant on the roadmap, so I'm OK with it given that it's kind of an edge case.

Reply:
@vito I have seen the contributor instruction for how to run `watsjs` test, but `yarn test` is just stuck. I don't know what is wrong with my environment.

Thanks for pointing out the root cause. To solve the core problem, I think we can just added a new boolean column `last_check_succeeded` to `resource_config_scopes`, so that we can easily how if last check succeeded. If last check failed, then it should run the check.

Even after solved the core problem, I still think we should distinguish regular build and check build in `checkDelegate.WaitToRun()`. Because `build.IsManuallyTriggered()` means different for regular build and check build.

For a check build, manual triggered check and lidar triggered check may behave differently.  But for a regular build, step embedded checks should behave in the same way no matter the build is triggered automatically or by manual.

The latest commit addressed the both problems. Please take a look again.